### PR TITLE
core: interrupt: update dt and client interface

### PR DESCRIPTION
**Pull request by ludovicbarre**
This patch series:
- Split dt-bindings management to welcome other interrupt drivers.
- Update gic driver to take account the type of interrupt specified in device tree (SPI | PPI)
- Add interrupt resource in dt_node_info.
- Add an interface to allocate and add an interrupt handler.

### CODE REVIEW

File: core/drivers/gic.c [hunk 1]
```diff
@@ -205,13 +207,28 @@ void gic_init(struct gic_data *gd, vaddr_t gicc_base __maybe_unused,
 #endif
 }
 
+static int __maybe_unused gic_xlate(const uint32_t *properties, int len)
+{
+	int it_num = DT_INFO_INVALID_INTERRUPT;
```
Comment (jenswi-linaro) [inline core/drivers/gic.c:212]: This function can be simplified to
``` C
if (!properties || len < 2)
      return DT_INFO_INVALID_INTERRUPT;

return fdt32_to_cpu(properties[1]);

**Review by jenswi-linaro — COMMENTED** [review]
Comment for the commit "core: dt: split dt_get_irq() between interrupt framework and drivers"

File: core/drivers/gic.c (hunk 1 above)
Comment (ludovicbarre) [reply core/drivers/gic.c:212]: Ok no problem

File: core/include/kernel/interrupt.h [hunk 2]
```diff
@@ -42,6 +43,18 @@ struct itr_handler {
 
 void itr_init(struct itr_chip *data);
 void itr_handle(size_t it);
+#ifdef CFG_DT
```
Comment (etienne-lms) [inline core/include/kernel/interrupt.h:50]: add an empty line above

File: core/include/kernel/interrupt.h [hunk 3]
```diff
@@ -42,6 +43,18 @@ struct itr_handler {
...
+/*
+ * Get the DT interrupt property of the @node. In the DT an interrupt
```
Comment (etienne-lms) [inline core/include/kernel/interrupt.h:52]: needs rephrasing, maybe simply:
` * Get the DT interrupt property at @node`

**Review by etienne-lms — COMMENTED** [review]
2 minor comments for commit "core: dt: split dt_get_irq() between interrupt framework and drivers", otherwise `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for all.

Regarding `itr_free()` i don't agree with @jforissier: I think we should implement the function, for consistency, despite it may not be used in currently supported platforms.

File: core/drivers/gic.c [hunk 4]
```diff
@@ -205,13 +207,28 @@ void gic_init(struct gic_data *gd, vaddr_t gicc_base __maybe_unused,
...
+}
+
 void gic_init_base_addr(struct gic_data *gd, vaddr_t gicc_base __maybe_unused,
 			vaddr_t gicd_base)
 {
 	gd->gicc_base = gicc_base;
 	gd->gicd_base = gicd_base;
 	gd->max_it = probe_max_it(gicc_base, gicd_base);
 	gd->chip.ops = &gic_ops;
+#ifdef CFG_DT
```
Comment (jforissier) [inline core/drivers/gic.c:240]: `if IS_ENABLED(CFG_DT)` and remove `__maybe_unused` above.

File: core/include/kernel/interrupt.h [hunk 5]
```diff
@@ -14,6 +14,7 @@
 
 struct itr_chip {
 	const struct itr_ops *ops;
+	int (*xlate)(const uint32_t *properties, int len);
```
Comment (jforissier) [inline core/include/kernel/interrupt.h:17]: Please find a more meaningful name. How about `int (*dt_get_irq)(const uint32_t *prop, int len);`?

File: core/kernel/interrupt.c [hunk 6]
```diff
@@ -65,6 +66,33 @@ void itr_handle(size_t it)
 	}
 }
 
+struct itr_handler *itr_alloc_add(size_t it, itr_handler_t handler,
+				  uint32_t flags, void *data)
+{
+	struct itr_handler *hdl = calloc(1, sizeof(struct itr_handler));
```
Comment (jforissier) [inline core/kernel/interrupt.c:72]: `sizeof(*hdl)`

**Review by jforissier — COMMENTED** [review]
- For commit "core: dt: split dt_get_irq() between interrupt framework and drivers":
In the commit description: s/gic/GIC/, s/arm/Arm/, s/specially/specifically/
With my comments addressed:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`
- For commit "core: dt: take account type of interrupt in dt_get_irq()":
Please use proper capitalization in commit description: s/caam/CAAM/.
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`
- For commit "core: add interrupt resource in dt_node_info":
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`
- For commit " core: kernel: interrupt: add interface to allocate and add handler":
One comment below, then:
`Reviewed-by: Jerome Forissier <jerome@forissier.org>`

**Comment by ludovicbarre** [comment]
thx for your review, I add tags review and squash the modifications.

File: core/drivers/gic.c [hunk 7]
```diff
@@ -210,10 +210,25 @@ void gic_init(struct gic_data *gd, vaddr_t gicc_base __maybe_unused,
 
 static int gic_dt_get_irq(const uint32_t *properties, int len)
 {
+	int it_num = DT_INFO_INVALID_INTERRUPT;
+
 	if (!properties || len < 2)
-		return DT_INFO_INVALID_INTERRUPT;
+		return it_num;
```
Comment (etienne-lms) [inline core/drivers/gic.c:216]: minor: would prefer explicit `return DT_INFO_INVALID_INTERRUPT;` here.

**Comment by ludovicbarre** [comment]
re-send with etienne comment

**Review by etienne-lms — APPROVED** [review]
LGTM.
Note commit "core: kernel: interrupt: add interface to allocate and add handler" implements `itr_free()`. Jérôme, your tag is applied, are you ok with this?

**Comment by jforissier** [comment]
@etienne-lms yes that's fine.

**Comment by jforissier** [comment]
Azure remained stuck for more than 20 minutes when pulling the Docker image and as a result it timed out. I have restarted it and will merge this as soon as it completes successfully.

**Pull request merged**